# Cleaning and exploring data for Macro-Eyes

In [201]:
import os
import pathlib
from pathlib import Path 
import pandas as pd
import numpy as np
import re

To do:
* Remove duplicate rows
* Check for different text referring to same facility eg 'hospital' vs 'chp'

It will be helpful later to display all the columns:

In [202]:
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 200)

## Read in raw data

In [203]:
df_raw = pd.read_excel('Health Facility Assessment (1).xlsx',index_col=0, usecols='B:GW')
df = df_raw.copy()

## Delete empty columns

In [204]:
# the columns whose names end with 'choice_labels' are empty, so we can delete them
choice_label_cols = []
for col in df.columns:
    if len((re.findall('\.choice_labels',col)))!=0:
        choice_label_cols.append(col)

df = df.drop(choice_label_cols,axis=1)


## Shorten column names

Let's start by displaying the column names:

In [205]:
cols = list(df.columns)
print(*cols,sep='\n')

form.health_centre_information.facility_name
form.facility_gps
form.health_centre_information.location_information.region_province
form.health_centre_information.location_information.district
form.health_centre_information.location_information.chiefdom
form.health_centre_information.location_information.facility_location
form.health_centre_information.facility_type
form.health_centre_information.facility_type_other
form.health_centre_information.managing_authority
form.health_centre_information.managing_authority_other
form.health_centre_information.setting
form.health_centre_information.outpatient_only
form.health_centre_information.capacity.number_consultation_rooms
form.health_centre_information.capacity.number_inpatient_beds
form.health_centre_information.capacity.number_maternity_beds
form.health_centre_information.group_number_employed.list_community_health_officer.cho_number_in_post
form.health_centre_information.group_number_employed.list_community_health_officer.cho_number_pre

These column names are painfully long to work with, so I will remove some unnecesary words. 

In [206]:
new_cols = []
for i,name in enumerate(cols):
    name = re.sub('form\.','',name)
    name = re.sub('health_centre_information\.','',name)
    name = re.sub('location_information\.','',name)
    name = re.sub('group_number_employed\.','',name)
    name = re.sub('list_.*\.','',name)
    name = re.sub('ql_information_education_communication\.','',name)
    name = re.sub('ql_human_resources\.','',name)
    name = re.sub('ql_surveillance\.','',name)
    name = re.sub('ql_triage_and_early_recognition\.','',name)
    name = re.sub('ql_chw\.','',name)
    name = re.sub('ql_isolation_physical_distancing\.','',name)
    
    # remove everything from 'grp_infection_prevention_and_control.' to the next '.'
    name = re.sub('grp_infection_prevention_and_control\..*\.','',name) 
    # remove any remaining instances of 'grp_infection_prevention_and_control.'
    name = re.sub('grp_infection_prevention_and_control\.','',name) 
    
    # remove everything before the word 'log'
    name = re.sub('.*\.log_','log_',name)
    
    new_cols.append(name)

    print(name)
df.columns = new_cols

facility_name
facility_gps
region_province
district
chiefdom
facility_location
facility_type
facility_type_other
managing_authority
managing_authority_other
setting
outpatient_only
capacity.number_consultation_rooms
capacity.number_inpatient_beds
capacity.number_maternity_beds
cho_number_in_post
cho_number_present_at_visit
cht_number_in_post
cht_number_present_at_visit
cha_number_in_post
cha_number_present_at_visit
cm_number_in_post
cm_number_present_at_visit
sechn_number_in_post
sechn_number_present_at_visit
lab_tech_number_in_post
lab_tech_number_present_at_visit
mch_aides_number_in_post
mch_aides_number_present_at_visit
cleaner_porter_number_in_post
cleaner_porter_number_present_at_visit
other_number_in_post
other_number_present_at_visit
other_staff_cadre
sections_to_review
consultations.head_count.month_1
consultations.head_count.month_2
consultations.head_count.month_3
consultations.head_count.month_4
consultations.general_outpatient.month_1
consultations.general_outpatient.month_

Abbreviations in the new column names:
* cho: community health officer
* cht: community health technician
* cha: community health assistant
* cm: community midwives
* sechn: state enrolled community health nurse
* mch: maternal and child health
* hr: human resources
* iec: information education communication
* surv: surveillance
* ter: triage and early recognition
* chw: community health worker
* iso: isolation
* ppe: personal protective equipment
* ipc: infection prevention and control
* wcd: waste collection and disposal
* ds: disinfection and sterilization
* log: logistics

## Some preliminary changes to the data

There are a few obvious things we can do right off the bat to make the data easier to work with. 

Firstly, the missing values in the dataset are indicated by '---'. Explicitly saying that the values are missing is more useful, so we replace the '---' values with np.nan. 

In [207]:
# replace --- values with NA
df = df.replace('---', np.nan, regex=True)

We can also split the GPS values into latitude and longitude:

In [208]:
# split GPS values into longitude and latitude
df.insert(loc=1,column = 'facility_gps_lat',value=df['facility_gps'].apply(lambda x: x.split(',')[0]))
df.insert(loc=2,column = 'facility_gps_lon',value=df['facility_gps'].apply(lambda x: x.split(',')[1]))
df = df.drop('facility_gps',axis=1)

/tmp/ipykernel_23591/4057685809.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df.insert(loc=1,column = 'facility_gps_lat',value=df['facility_gps'].apply(lambda x: x.split(',')[0]))
/tmp/ipykernel_23591/4057685809.py:3: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df.insert(loc=2,column = 'facility_gps_lon',value=df['facility_gps'].apply(lambda x: x.split(',')[1]))


We can change yes and no values to true and false values:

In [209]:
# replace 'yes' with True and 'no' with False
df['outpatient_only'] = df['outpatient_only'].apply(lambda x: True if x=='yes' else x)
df['outpatient_only'] = df['outpatient_only'].apply(lambda x: False if x=='no' else x)


## Data types

The dataframe will be easier to use for modeling and to work with in general if the data types for the columns are specified. Let us explore the current data types:

In [210]:
df.dtypes

facility_name                                                                 object
facility_gps_lat                                                              object
facility_gps_lon                                                              object
region_province                                                               object
district                                                                      object
chiefdom                                                                      object
facility_location                                                             object
facility_type                                                                 object
facility_type_other                                                          float64
managing_authority                                                            object
managing_authority_other                                                      object
setting                                                          

There are lots of 'object' data types. Let's make the dataframe easier to use for modelling by assigning the correct data types. There are a lot of columns, so I assign data types in a bit of a hacky way below to save time. I consider the part of the column name after the last full stop. If it contains a particular keyword, I give it a particular data type. I use 16 bits for the integers because it's very unlikely that any facility will have more than around 32000 of any of the quantities in this database. 

In [211]:
type_dict = {} # make a dictionary specifying data types for columns 
for col in df.columns:
    name_parts = col.split('.')
    
    if len(re.findall('number_', name_parts[-1]))!=0:
        type_dict[col] = 'Int16'
    elif name_parts[-1][:6]=='month_':
        type_dict[col] = 'Int16'
    elif name_parts[-1][:3]=='hr_':
        type_dict[col] = 'string'
    elif name_parts[-1][:4]=='iec_':
        type_dict[col] = 'string'
    elif name_parts[-1][:5]=='surv_':
        type_dict[col] = 'string'
    elif name_parts[-1][:4]=='ter_':
        type_dict[col] = 'string'
    elif name_parts[-1][:4]=='chw_':
        type_dict[col] = 'string'
    elif name_parts[-1][:4]=='iso_':
        type_dict[col] = 'string'
    elif name_parts[-1][:4]=='ipc_':
        type_dict[col] = 'string'
    elif name_parts[-1][:4]=='log_':
        type_dict[col] = 'string'
    elif name_parts[-1][:15]=='monthly_average':
        type_dict[col] = 'float64'
    elif name_parts[-1][:6]=='score_':
        type_dict[col] = 'float64'
type_dict_2 = {'facility_name': 'string', 'facility_gps_lat': 'float64', 'facility_gps_lon':'float64', 'region_province': 'string', 'district': 'string', 'chiefdom': 'string', 'facility_location': 'string', 'facility_type': 'string', 'facility_type_other': 'string', 'managing_authority': 'string', 'managing_authority_other': 'string', 'setting': 'string', 'outpatient_only': 'boolean' }
type_dict = type_dict | type_dict_2 # join two dictionaries

df=df.astype(type_dict)

We can check that the data types are now correct:

In [212]:
df.dtypes

facility_name                                                                 string
facility_gps_lat                                                             float64
facility_gps_lon                                                             float64
region_province                                                               string
district                                                                      string
chiefdom                                                                      string
facility_location                                                             string
facility_type                                                                 string
facility_type_other                                                           string
managing_authority                                                            string
managing_authority_other                                                      string
setting                                                          

## Specifying case

Python and many other languages are case sensitive, so a word like 'Python' is treated differently from 'python'. We can better detect duplicates if we make the case consistent throughout. 

In [213]:
# get the names of all the string columns
str_cols = [i for i in df.columns if df.dtypes[i]=='string']

# make the string columns all lowercase
df[str_cols] = df[str_cols].apply(lambda x: x.str.lower(),axis=1)

# make location names all uppercase
df['facility_name'] = df['facility_name'].str.upper()
df['region_province'] = df['region_province'].str.upper()
df['district'] = df['district'].str.upper()
df['chiefdom'] = df['chiefdom'].str.upper()
df['facility_location'] = df['facility_location'].str.upper()


## Dealing with duplicates

Now it's time to investigate duplicate data. 

Let's look for duplicate facility names:

In [214]:
df_dup_1 = df[df.duplicated(['facility_name'],keep=False)].sort_values('facility_name')
df_dup_1

facility_name  facility_gps_lat  facility_gps_lon  \
number                                                                 
44                     AMBER CHP        -51.206867         -4.695275   
41                     AMBER CHP        -63.524311          4.306230   
87                       AMY CHP         70.462361        122.083912   
129                      AMY CHP         73.580693        -15.861572   
48                       AMY CHP         84.498725         48.271191   
24                     DAVID CHP         32.649001          0.236277   
132                    DAVID CHP         46.939023        127.075406   
52                     DAVID CHP         42.120740        -86.197466   
110                   GEORGE CHC        -39.737294       -109.387845   
56                    GEORGE CHC        -85.563514       -123.262201   
25                 JOANNA CLINIC         15.754015       -153.871776   
47                 JOANNA CLINIC         60.190934        162.787117   
95                   JOHN CLINIC         53.647814       -115.559792   
26                   JOHN CLINIC        -54.388258       -131.023446   
107                  JOHN CLINIC         47.128242         54.632433   
39                KEVIN HOSPITAL        -53.597735        136.778861   
133               KEVIN HOSPITAL         41.570610         66.484104   
57        LISA TERTIARY HOSPITAL          7.736792         41.847715   
80        LISA TERTIARY HOSPITAL         50.873618       -160.070770   
113               MELISSA CLINIC        -29.047545       -115.874086   
86                MELISSA CLINIC         35.576733        179.184687   
141              NICOLE HOSPITAL        -56.293473       -143.818347   
115              NICOLE HOSPITAL         -1.046347        153.958550   
140     NICOLE TERTIARY HOSPITAL         63.155728        -82.989076   
79      NICOLE TERTIARY HOSPITAL        -41.344766       -134.396481   
96                  RICHARD MCHP        -26.341916       -172.661039   
97                  RICHARD MCHP        -81.332628         69.862048   
122                 RICHARD MCHP         28.935312        -68.439807   
16               ROBERT HOSPITAL         43.308952        134.454731   
83               ROBERT HOSPITAL         -6.673749       -118.231429   
82      ROBERT REGIONAL HOSPITAL         31.051884        119.526785   
42      ROBERT REGIONAL HOSPITAL        -15.224350         87.259650   
128                SEAN HOSPITAL         11.498693        168.848917   
101                SEAN HOSPITAL         81.489324        -62.510515   
125                    TAMMY CHC        -11.824177         84.379063   
4                      TAMMY CHC         54.714883        150.744813   
78               THOMAS HOSPITAL        -21.221849        -47.055607   
50               THOMAS HOSPITAL         81.553500        -10.155364   
5                           <NA>         22.856753         19.262410   
119                         <NA>         -1.129569        -34.179477   

       region_province            district        chiefdom  facility_location  \
number                                                                          
44            SOUTHERN             MOYAMBA         ALABAMA     NEW WALTERPORT   
41       NORTH-WESTERN              KARENE   MASSACHUSETTS    PORT CAROLBURGH   
87             WESTERN  WESTERN AREA URBAN       WISCONSIN     EAST JAMESBERG   
129            WESTERN  WESTERN AREA URBAN    RHODE ISLAND      WEST LOUISTON   
48             WESTERN  WESTERN AREA URBAN       WISCONSIN     EAST JAMESBERG   
24        WESTERN AREA  WESTERN AREA RURAL       LOUISIANA           MARKVIEW   
132           SOUTHERN             MOYAMBA    PENNSYLVANIA         RHONDATOWN   
52             EASTERN            KAILAHUN          OREGON          SOUTH AMY   
110            EASTERN              KENEMA           TEXAS         PHILIPLAND   
56             EASTERN              KENEMA           TEXAS         PHILIPLAND   
25            SOUTHERN

Let's see how many elements are in this dataframe:

In [215]:
len(df_dup_1)

40

It is possible that some of the duplicate facility names refer to independent clinics that happen to have the same name. Let's also consider the number of consultations in the facility to get a clearer picture. 

In [216]:
df_dup_2 = df[df.duplicated(['facility_name','consultations.head_count.month_1','consultations.head_count.month_2','consultations.head_count.month_3','consultations.head_count.month_4'],keep=False)].sort_values('facility_name')
df_dup_2

facility_name  facility_gps_lat  facility_gps_lon  \
number                                                                 
48                       AMY CHP         84.498725         48.271191   
87                       AMY CHP         70.462361        122.083912   
56                    GEORGE CHC        -85.563514       -123.262201   
110                   GEORGE CHC        -39.737294       -109.387845   
25                 JOANNA CLINIC         15.754015       -153.871776   
47                 JOANNA CLINIC         60.190934        162.787117   
86                MELISSA CLINIC         35.576733        179.184687   
113               MELISSA CLINIC        -29.047545       -115.874086   
115              NICOLE HOSPITAL         -1.046347        153.958550   
141              NICOLE HOSPITAL        -56.293473       -143.818347   
79      NICOLE TERTIARY HOSPITAL        -41.344766       -134.396481   
140     NICOLE TERTIARY HOSPITAL         63.155728        -82.989076   
83               ROBERT HOSPITAL         -6.673749       -118.231429   
16               ROBERT HOSPITAL         43.308952        134.454731   
101                SEAN HOSPITAL         81.489324        -62.510515   
128                SEAN HOSPITAL         11.498693        168.848917   
125                    TAMMY CHC        -11.824177         84.379063   
4                      TAMMY CHC         54.714883        150.744813   

       region_province            district      chiefdom facility_location  \
number                                                                       
48             WESTERN  WESTERN AREA URBAN     WISCONSIN    EAST JAMESBERG   
87             WESTERN  WESTERN AREA URBAN     WISCONSIN    EAST JAMESBERG   
56             EASTERN              KENEMA         TEXAS        PHILIPLAND   
110            EASTERN              KENEMA         TEXAS        PHILIPLAND   
25            SOUTHERN                  BO   MISSISSIPPI      AMBERCHESTER   
47            SOUTHERN                  BO   MISSISSIPPI      AMBERCHESTER   
86            NORTHERN           KOINADUGU     TENNESSEE     ANDREACHESTER   
113           NORTHERN           KOINADUGU     TENNESSEE     ANDREACHESTER   
115           SOUTHERN                  BO      VIRGINIA      STEVECHESTER   
141           SOUTHERN                  BO      VIRGINIA      STEVECHESTER   
79            SOUTHERN                  BO  RHODE ISLAND     NEW CHRISTINE   
140           SOUTHERN                  BO  RHODE ISLAND     NEW CHRISTINE   
83       NORTH-WESTERN           PORT LOKO      KENTUCKY    VALERIEBOROUGH   
16       NORTH-WESTERN           PORT LOKO      KENTUCKY    VALERIEBOROUGH   
101           SOUTHERN                  BO      NEW YORK        CLARKMOUTH   
128           SOUTHERN                  BO      NEW YORK         GUZMANTON   
125           SOUTHERN              BONTHE      VIRGINIA          RENEETON   
4             SOUTHERN              BONTHE      VIRGINIA          RENEETON   

                  facility_type facility_type_other managing_authority  \
number                                                                   
48        community_health_post                <NA>  government_public   
87        community_health_post                <NA>  government_public   
56      community_health_centre                <NA>  government_public   
110     community_health_centre                <NA>  government_public   
25                     mch_post                <NA>  government_public   
47                     mch_post                <NA>  government_public   
86        community_health_post                <NA>  government_public   
113       community_health_post                <NA>  government_public   
115       community_health_post                <NA>  government_public   
141       community_health_post                <NA>  government_public   
79      community_health_centre                <NA>  government_public   
140     community_health_centre                <NA>  go

It's very unlikely that the consultation headcount numbers for four months would be exactly the same between two identically named clinics. These are therefore likely to truly refer to the same facility.



In [217]:
len(df_dup_2)

18

Our number of duplicates has now dropped lower. But maybe there are still some very similar entries where one or more of the consultation headcounts differ. We have not recorded those in df_dup_2. Let's search for matches with regard to a number of different column combinations and merge the resulting dataframes into one. 

We consider matching facility names and facility locations first.

In [218]:
df_dup_3 = df[df.duplicated(['facility_name','facility_location'],keep=False)].sort_values('facility_name')
df_dup_3

facility_name  facility_gps_lat  facility_gps_lon  \
number                                                                 
48                       AMY CHP         84.498725         48.271191   
87                       AMY CHP         70.462361        122.083912   
56                    GEORGE CHC        -85.563514       -123.262201   
110                   GEORGE CHC        -39.737294       -109.387845   
25                 JOANNA CLINIC         15.754015       -153.871776   
47                 JOANNA CLINIC         60.190934        162.787117   
26                   JOHN CLINIC        -54.388258       -131.023446   
95                   JOHN CLINIC         53.647814       -115.559792   
133               KEVIN HOSPITAL         41.570610         66.484104   
39                KEVIN HOSPITAL        -53.597735        136.778861   
86                MELISSA CLINIC         35.576733        179.184687   
113               MELISSA CLINIC        -29.047545       -115.874086   
115              NICOLE HOSPITAL         -1.046347        153.958550   
141              NICOLE HOSPITAL        -56.293473       -143.818347   
140     NICOLE TERTIARY HOSPITAL         63.155728        -82.989076   
79      NICOLE TERTIARY HOSPITAL        -41.344766       -134.396481   
97                  RICHARD MCHP        -81.332628         69.862048   
96                  RICHARD MCHP        -26.341916       -172.661039   
83               ROBERT HOSPITAL         -6.673749       -118.231429   
16               ROBERT HOSPITAL         43.308952        134.454731   
125                    TAMMY CHC        -11.824177         84.379063   
4                      TAMMY CHC         54.714883        150.744813   
5                           <NA>         22.856753         19.262410   
119                         <NA>         -1.129569        -34.179477   

       region_province            district       chiefdom facility_location  \
number                                                                        
48             WESTERN  WESTERN AREA URBAN      WISCONSIN    EAST JAMESBERG   
87             WESTERN  WESTERN AREA URBAN      WISCONSIN    EAST JAMESBERG   
56             EASTERN              KENEMA          TEXAS        PHILIPLAND   
110            EASTERN              KENEMA          TEXAS        PHILIPLAND   
25            SOUTHERN                  BO    MISSISSIPPI      AMBERCHESTER   
47            SOUTHERN                  BO    MISSISSIPPI      AMBERCHESTER   
26            NORTHERN           KOINADUGU      TENNESSEE          GARYTOWN   
95            NORTHERN           KOINADUGU      TENNESSEE          GARYTOWN   
133      NORTH-WESTERN              KARENE  MASSACHUSETTS        HUNTERPORT   
39       NORTH-WESTERN              KARENE  MASSACHUSETTS        HUNTERPORT   
86            NORTHERN           KOINADUGU      TENNESSEE     ANDREACHESTER   
113           NORTHERN           KOINADUGU      TENNESSEE     ANDREACHESTER   
115           SOUTHERN                  BO       VIRGINIA      STEVECHESTER   
141           SOUTHERN                  BO       VIRGINIA      STEVECHESTER   
140           SOUTHERN                  BO   RHODE ISLAND     NEW CHRISTINE   
79            SOUTHERN                  BO   RHODE ISLAND     NEW CHRISTINE   
97             WESTERN  WESTERN AREA URBAN       KENTUCKY      COLLEENBURGH   
96             WESTERN  WESTERN AREA URBAN       KENTUCKY      COLLEENBURGH   
83       NORTH-WESTERN           PORT LOKO       KENTUCKY    VALERIEBOROUGH   
16       NORTH-WESTERN           PORT LOKO       KENTUCKY    VALERIEBOROUGH   
125           SOUTHERN              BONTHE       VIRGINIA          RENEETON   
4             SOUTHERN              BONTHE       VIRGINIA          RENEETON   
5                 <NA>                <NA>           <NA>              <NA>   
119               <NA>                <NA>           <NA>              <NA>   

                  facility_type facility_type_other managing_authority  \
number           

In [219]:
len(df_dup_3)

24

Now we consider cases where at least one of the months' consultation headcounts match. This seems unlikely enough that it's worth looking into.

In [220]:
# considering rows where consultation headcount month 1 matches
df_dup_4 = df[df.duplicated(['facility_name','consultations.head_count.month_1'],keep=False)].sort_values('facility_name')
# considering rows where consultation headcount month 2 matches
df_dup_5 = df[df.duplicated(['facility_name','consultations.head_count.month_2'],keep=False)].sort_values('facility_name')
# considering rows where consultation headcount month 3 matches
df_dup_6 = df[df.duplicated(['facility_name','consultations.head_count.month_3'],keep=False)].sort_values('facility_name')
# considering rows where consultation headcount month 4 matches
df_dup_7 = df[df.duplicated(['facility_name','consultations.head_count.month_4'],keep=False)].sort_values('facility_name')

There will be a lot of overlap among df_dup_3, df_dup_4, df_dup_5, df_dup_6, and df_dup_7. We want to save all the unique entries across these dataframes, without having extra duplicates. We start by merging the duplicate candidates into one dataframe, in the process creating lots of duplicates. Those duplicates need to be deleted later.  

In [221]:
df_dup_potential_redundant = pd.concat([df_dup_3,df_dup_4,df_dup_5,df_dup_6,df_dup_7])

We now find and delete the duplicate indices. 

In [225]:
df_dup_potential = df_dup_potential_redundant[~df_dup_potential_redundant.index.duplicated(keep='first')]
#dup_indices = df_dup_potential_redundant.index.duplicated(keep='first')
#df_dup_potential = df_dup_potential_redundant.drop(df_dup_potential_redundant[dup_indices].index)
df_dup_potential

facility_name  facility_gps_lat  facility_gps_lon  \
number                                                                 
48                       AMY CHP         84.498725         48.271191   
87                       AMY CHP         70.462361        122.083912   
56                    GEORGE CHC        -85.563514       -123.262201   
110                   GEORGE CHC        -39.737294       -109.387845   
25                 JOANNA CLINIC         15.754015       -153.871776   
47                 JOANNA CLINIC         60.190934        162.787117   
26                   JOHN CLINIC        -54.388258       -131.023446   
95                   JOHN CLINIC         53.647814       -115.559792   
133               KEVIN HOSPITAL         41.570610         66.484104   
39                KEVIN HOSPITAL        -53.597735        136.778861   
86                MELISSA CLINIC         35.576733        179.184687   
113               MELISSA CLINIC        -29.047545       -115.874086   
115              NICOLE HOSPITAL         -1.046347        153.958550   
141              NICOLE HOSPITAL        -56.293473       -143.818347   
140     NICOLE TERTIARY HOSPITAL         63.155728        -82.989076   
79      NICOLE TERTIARY HOSPITAL        -41.344766       -134.396481   
97                  RICHARD MCHP        -81.332628         69.862048   
96                  RICHARD MCHP        -26.341916       -172.661039   
83               ROBERT HOSPITAL         -6.673749       -118.231429   
16               ROBERT HOSPITAL         43.308952        134.454731   
125                    TAMMY CHC        -11.824177         84.379063   
4                      TAMMY CHC         54.714883        150.744813   
5                           <NA>         22.856753         19.262410   
119                         <NA>         -1.129569        -34.179477   
101                SEAN HOSPITAL         81.489324        -62.510515   
128                SEAN HOSPITAL         11.498693        168.848917   

       region_province            district       chiefdom facility_location  \
number                                                                        
48             WESTERN  WESTERN AREA URBAN      WISCONSIN    EAST JAMESBERG   
87             WESTERN  WESTERN AREA URBAN      WISCONSIN    EAST JAMESBERG   
56             EASTERN              KENEMA          TEXAS        PHILIPLAND   
110            EASTERN              KENEMA          TEXAS        PHILIPLAND   
25            SOUTHERN                  BO    MISSISSIPPI      AMBERCHESTER   
47            SOUTHERN                  BO    MISSISSIPPI      AMBERCHESTER   
26            NORTHERN           KOINADUGU      TENNESSEE          GARYTOWN   
95            NORTHERN           KOINADUGU      TENNESSEE          GARYTOWN   
133      NORTH-WESTERN              KARENE  MASSACHUSETTS        HUNTERPORT   
39       NORTH-WESTERN              KARENE  MASSACHUSETTS        HUNTERPORT   
86            NORTHERN           KOINADUGU      TENNESSEE     ANDREACHESTER   
113           NORTHERN           KOINADUGU      TENNESSEE     ANDREACHESTER   
115           SOUTHERN                  BO       VIRGINIA      STEVECHESTER   
141           SOUTHERN                  BO       VIRGINIA      STEVECHESTER   
140           SOUTHERN                  BO   RHODE ISLAND     NEW CHRISTINE   
79            SOUTHERN                  BO   RHODE ISLAND     NEW CHRISTINE   
97             WESTERN  WESTERN AREA URBAN       KENTUCKY      COLLEENBURGH   
96             WESTERN  WESTERN AREA URBAN       KENTUCKY      COLLEENBURGH   
83       NORTH-WESTERN           PORT LOKO       KENTUCKY    VALERIEBOROUGH   
16       NORTH-WESTERN           PORT LOKO       KENTUCKY    VALERIEBOROUGH   
125           SOUTHERN              BONTHE       VIRGINIA          RENEETON   
4             SOUTHERN              BONTHE       VIRGINIA          RENEETON   
5                 <NA>                <NA>           <NA>              <NA>   
119               <NA>     

I'm confident that the duplicates in df_dup_2 are genuine duplicates, so I'm going to remove them from df_dup_potential to make it easier to examine the more dubious cases. 

In [226]:
dup_2_indices = list(df_dup_2.index)

In [227]:
df_dup_potential_2 = df_dup_potential.drop(dup_2_indices,axis=0)

In [228]:
df_dup_potential_2

facility_name  facility_gps_lat  facility_gps_lon region_province  \
number                                                                       
26         JOHN CLINIC        -54.388258       -131.023446        NORTHERN   
95         JOHN CLINIC         53.647814       -115.559792        NORTHERN   
133     KEVIN HOSPITAL         41.570610         66.484104   NORTH-WESTERN   
39      KEVIN HOSPITAL        -53.597735        136.778861   NORTH-WESTERN   
97        RICHARD MCHP        -81.332628         69.862048         WESTERN   
96        RICHARD MCHP        -26.341916       -172.661039         WESTERN   
5                 <NA>         22.856753         19.262410            <NA>   
119               <NA>         -1.129569        -34.179477            <NA>   

                  district       chiefdom facility_location  \
number                                                        
26               KOINADUGU      TENNESSEE          GARYTOWN   
95               KOINADUGU      TENNESSEE          GARYTOWN   
133                 KARENE  MASSACHUSETTS        HUNTERPORT   
39                  KARENE  MASSACHUSETTS        HUNTERPORT   
97      WESTERN AREA URBAN       KENTUCKY      COLLEENBURGH   
96      WESTERN AREA URBAN       KENTUCKY      COLLEENBURGH   
5                     <NA>           <NA>              <NA>   
119                   <NA>           <NA>              <NA>   

                  facility_type facility_type_other managing_authority  \
number                                                                   
26                     mch_post                <NA>  government_public   
95                     mch_post                <NA>  government_public   
133     community_health_centre                <NA>  government_public   
39      community_health_centre                <NA>  government_public   
97        community_health_post                <NA>              other   
96        community_health_post                <NA>              other   
5                          <NA>                <NA>               <NA>   
119                        <NA>                <NA>               <NA>   

       managing_authority_other     setting  outpatient_only  \
number                                                         
26                         <NA>  peri_urban             True   
95                         <NA>  peri_urban             True   
133                        <NA>  peri_urban            False   
39                         <NA>  peri_urban            False   
97              private for fhc       urban             True   
96              private for fhc       urban             True   
5                          <NA>        <NA>             <NA>   
119                        <NA>        <NA>             <NA>   

        capacity.number_consultation_rooms  capacity.number_inpatient_beds  \
number                                                                       
26                                       0                               0   
95                                       1                               2   
133                                      3                              11   
39                                       3                               3   
97                                       1                               6   
96                                       1                               6   
5                                     <NA>                            <NA>   
119                                   <NA>                            <NA>   

        capacity.number_maternity_beds  cho_number_in_post  \
number                                                       
26                                   1                   2   
95                                   2                  15   
133                                  1                   0   
39                                   1                   1   
97                                   2                   0   


Now we're left with only four more potential duplicates to investigate.

For each potential match, let's make a mini dataframe.

In [250]:
display(df_dup_potential_2['facility_name'].isnull())

number
26     False
95     False
133    False
39     False
97     False
96     False
5       True
119     True
Name: facility_name, dtype: bool

In [272]:
mini_dfs = []
mini_df_dict = {}
clinic_names = df_dup_potential_2['facility_name'].unique()
for name in clinic_names:
    if not pd.isnull(name): # if the facility name is not null
        mini_df = df_dup_potential_2[df_dup_potential_2['facility_name']==name]
    else: # when the facility name is null   
        mini_df = df_dup_potential_2[df_dup_potential_2['facility_name'].isnull()]
    mini_df_dict[name]=mini_df
    mini_dfs.append(mini_df)

In [271]:
for i,mini_df in enumerate(mini_dfs):
    facility_name = mini_df['facility_name'].iloc[0]
    print(f'Facility name: {facility_name}')
    indices = list(mini_df.index)
    print('Where the rows differ: ')
    #display(~(pd.isnull(mini_df.loc[indices[0]]) & pd.isnull(mini_df.loc[indices[1]])))
    display(mini_df[mini_df.columns[(mini_df.loc[indices[0]] != mini_df.loc[indices[1]]) & (~(pd.isnull(mini_df.loc[indices[0]]) & pd.isnull(mini_df.loc[indices[1]])))]])
    print('Where the rows are the same: ')
    display(mini_df[mini_df.columns[(mini_df.loc[indices[0]] == mini_df.loc[indices[1]]) | (pd.isnull(mini_df.loc[indices[0]]) & pd.isnull(mini_df.loc[indices[1]]))]])

Facility name: JOHN CLINIC
Where the rows differ: 


facility_gps_lat  facility_gps_lon  \
number                                       
26            -54.388258       -131.023446   
95             53.647814       -115.559792   

        capacity.number_consultation_rooms  capacity.number_inpatient_beds  \
number                                                                       
26                                       0                               0   
95                                       1                               2   

        capacity.number_maternity_beds  cho_number_in_post  \
number                                                       
26                                   1                   2   
95                                   2                  15   

        cho_number_present_at_visit  cht_number_in_post  \
number                                                    
26                                1                   0   
95                               12                   4   

        cht_number_present_at_visit  cm_number_in_post  \
number                                                   
26                                0                  0   
95                                2                  3   

        cm_number_present_at_visit  sechn_number_in_post  \
number                                                     
26                               0                     0   
95                               2                     5   

        sechn_number_present_at_visit  mch_aides_number_in_post  \
number                                                            
26                                  0                         2   
95                                  3                         5   

        mch_aides_number_present_at_visit  cleaner_porter_number_in_post  \
number                                                                     
26                                      1                              1   
95                                      4                              3   

        cleaner_porter_number_present_at_visit  other_number_in_post  \
number                                                                 
26                                           1                     3   
95                                           2                     0   

        other_number_present_at_visit other_staff_cadre  \
number                                                    
26                                  3               TBA   
95                                  0               NaN   

                                       sections_to_review  \
number                                                      
26      consultations hr information_education_communi...   
95                                          consultations   

        consultations.head_count.month_1  consultations.head_count.month_2  \
number                                                                       
26                                   340                               327   
95                                    14                                25   

        consultations.head_count.month_3  consultations.head_count.month_4  \
number                                                                       
26                                   368                               364   
95                                    10                                11   

        consultations.general_outpatient.month_1  \
number                                             
26                                           272   
95                                            12   

        consultations.general_outpatient.month_2  \
number                                             
26                                           267   
95                                             8   

        consultations.general_outpatient.month_3  \
number                                             
26                                           272   
95   

Where the rows are the same: 


facility_name region_province   district   chiefdom facility_location  \
number                                                                         
26       JOHN CLINIC        NORTHERN  KOINADUGU  TENNESSEE          GARYTOWN   
95       JOHN CLINIC        NORTHERN  KOINADUGU  TENNESSEE          GARYTOWN   

       facility_type facility_type_other managing_authority  \
number                                                        
26          mch_post                <NA>  government_public   
95          mch_post                <NA>  government_public   

       managing_authority_other     setting  outpatient_only  \
number                                                         
26                         <NA>  peri_urban             True   
95                         <NA>  peri_urban             True   

        cha_number_in_post  cha_number_present_at_visit  \
number                                                    
26                       0                            0   
95                       0                            0   

        lab_tech_number_in_post  lab_tech_number_present_at_visit  \
number                                                              
26                            0                                 0   
95                            0                                 0   

       ipc_ppe_medical_masks ipc_ppe_disp_surgical_masks  \
number                                                     
26                      <NA>                        <NA>   
95                      <NA>                        <NA>   

       ipc_ppe_eye_protection ipc_ppe_examination_gloves  \
number                                                     
26                       <NA>                       <NA>   
95                       <NA>                       <NA>   

       ipc_ppe_surgical_gloves ipc_ppe_long_cuffed_gloves  \
number                                                      
26                        <NA>                       <NA>   
95                        <NA>                       <NA>   

       ipc_ppe_heavy_duty_gloves ipc_ppe_long_sleeved_gown  \
number                                                       
26                          <NA>                      <NA>   
95                          <NA>                      <NA>   

       ipc_ppe_waterproof_aprons ipc_ppe_plan_staff_trained_on_ppe  \
number                                                               
26                          <NA>                              <NA>   
95                          <NA>                              <NA>   

       ipc_ppe_plan_ppe_poster_displayed ipc_ppe_plan_fit_test_kit  \
number                                                               
26                                  <NA>                      <NA>   
95                                  <NA>                      <NA>   

       ipc_ppe_plan_contingency_for_shortages ipc_wcd_colour_coded_bins  \
number                                                                    
26                                       <NA>                      <NA>   
95                                       <NA>                      <NA>   

       ipc_wcd_clinical_waste_bags  \
number                               
26                            <NA>   
95                            <NA>   

       ipc_wcd_laundry_receptacles_at_patient_rooms ipc_wcd_incinerator  \
number                                                                    
26                                             <NA>                <NA>   
95                                             <NA>                <NA>   

       ipc_wash_clean_running_water ipc_wash_hand_soap ipc_wash_liquid_soap  \
number                                                                        
26                             <NA>               <NA>                 <NA>   
95                             <NA>               <NA>                 <NA>   

       ipc_wash_disp_hand_towels ipc_wash_alcohol_ba

Facility name: KEVIN HOSPITAL
Where the rows differ: 


facility_gps_lat  facility_gps_lon  capacity.number_inpatient_beds  \
number                                                                       
133            41.570610         66.484104                              11   
39            -53.597735        136.778861                               3   

        cho_number_in_post  cm_number_present_at_visit  sechn_number_in_post  \
number                                                                         
133                      0                           0                     2   
39                       1                           1                     0   

        sechn_number_present_at_visit  consultations.head_count.month_1  \
number                                                                    
133                                 1                              1247   
39                                  0                              1840   

        consultations.head_count.month_2  consultations.head_count.month_3  \
number                                                                       
133                                  869                               781   
39                                  2330                              2309   

        consultations.head_count.month_4  \
number                                     
133                                  467   
39                                  1658   

        consultations.general_outpatient.month_1  \
number                                             
133                                          538   
39                                          1257   

        consultations.general_outpatient.month_2  \
number                                             
133                                          416   
39                                          1197   

        consultations.general_outpatient.month_3  \
number                                             
133                                          394   
39                                          1182   

        consultations.general_outpatient.month_4  \
number                                             
133                                          215   
39                                           803   

        consultations.deliveries.month_1  consultations.deliveries.month_2  \
number                                                                       
133                                   23                                22   
39                                    16                                16   

        consultations.deliveries.month_3  consultations.deliveries.month_4  \
number                                                                       
133                                   26                                 5   
39                                    16                                19   

        consultations.pent_vaccines.month_1  \
number                                        
133                                      23   
39                                       61   

        consultations.pent_vaccines.month_2  \
number                                        
133                                      28   
39                                       18   

        consultations.pent_vaccines.month_3  \
number                                        
133                                      17   
39                                       48   

        consultations.pent_vaccines.month_4 iec_when_facility_vs_home  \
number                                                                  
133                                      13                 displayed   
39                                       37             not_displayed   

       iec_helpline_number surv_procedure_for_notification  \
number                                                       
133              displayed               fully_operational   
39           not_displayed                    not_in_place   

       ter_s

Where the rows are the same: 


facility_name region_province district       chiefdom  \
number                                                           
133     KEVIN HOSPITAL   NORTH-WESTERN   KARENE  MASSACHUSETTS   
39      KEVIN HOSPITAL   NORTH-WESTERN   KARENE  MASSACHUSETTS   

       facility_location            facility_type facility_type_other  \
number                                                                  
133           HUNTERPORT  community_health_centre                <NA>   
39            HUNTERPORT  community_health_centre                <NA>   

       managing_authority managing_authority_other     setting  \
number                                                           
133     government_public                     <NA>  peri_urban   
39      government_public                     <NA>  peri_urban   

        outpatient_only  capacity.number_consultation_rooms  \
number                                                        
133               False                                   3   
39                False                                   3   

        capacity.number_maternity_beds  cho_number_present_at_visit  \
number                                                                
133                                  1                            0   
39                                   1                            0   

        cht_number_in_post  cht_number_present_at_visit  cha_number_in_post  \
number                                                                        
133                      0                            0                   0   
39                       0                            0                   0   

        cha_number_present_at_visit  cm_number_in_post  \
number                                                   
133                               0                  1   
39                                0                  1   

        lab_tech_number_in_post  lab_tech_number_present_at_visit  \
number                                                              
133                           0                                 0   
39                            0                                 0   

        mch_aides_number_in_post  mch_aides_number_present_at_visit  \
number                                                                
133                            0                                  0   
39                             0                                  0   

        cleaner_porter_number_in_post  cleaner_porter_number_present_at_visit  \
number                                                                          
133                                 0                                       0   
39                                  0                                       0   

        other_number_in_post  other_number_present_at_visit other_staff_cadre  \
number                                                                          
133                        0                              0               NaN   
39                         0                              0               NaN   

                                       sections_to_review hr_focal_point  \
number                                                                     
133     consultations hr information_education_communi...  not_completed   
39      consultations hr information_education_communi...  not_completed   

       hr_staff_received_info hr_healthcare_provider_training  \
number                                                          
133                 completed                   not_completed   
39                  completed                   not_completed   

       hr_healthcare_provider_revised_training hr_daily_staff_list  \
number                                                               
133                              not_completed           completed   
39                               not_completed           completed   

       iec_handwashing_procedure iec_ph

Facility name: RICHARD MCHP
Where the rows differ: 


facility_gps_lat  facility_gps_lon  cm_number_present_at_visit  \
number                                                                   
97            -81.332628         69.862048                           0   
96            -26.341916       -172.661039                           1   

                                       sections_to_review  \
number                                                      
97      consultations hr information_education_communi...   
96                                          consultations   

        consultations.head_count.month_1  \
number                                     
97                                  1301   
96                                  1204   

        consultations.general_outpatient.month_1 hr_focal_point  \
number                                                            
97                                           914  not_completed   
96                                           734           <NA>   

       hr_staff_received_info hr_healthcare_provider_training  \
number                                                          
97              not_completed                   not_completed   
96                       <NA>                            <NA>   

       hr_healthcare_provider_revised_training hr_daily_staff_list  \
number                                                               
97                               not_completed       not_completed   
96                                        <NA>                <NA>   

       iec_handwashing_procedure iec_physical_distancing  \
number                                                     
97                     displayed           not_displayed   
96                          <NA>                    <NA>   

       iec_covering_nose_mouth iec_early_symptom_recognition  \
number                                                         
97               not_displayed                 not_displayed   
96                        <NA>                          <NA>   

       iec_when_facility_vs_home iec_rational_ppe_use iec_helpline_number  \
number                                                                      
97                 not_displayed        not_displayed       not_displayed   
96                          <NA>                 <NA>                <NA>   

       surv_procedure_for_notification surv_official_case_definition  \
number                                                                 
97                        not_in_place                  not_in_place   
96                                <NA>                          <NA>   

       surv_hotline_number surv_timely_data_reported_to_district  \
number                                                             
97            not_in_place                     fully_operational   
96                    <NA>                                  <NA>   

       ter_screening_area_set_up ter_symptom_screening_questionnaires  \
number                                                                  
97                  not_in_place                         not_in_place   
96                          <NA>                                 <NA>   

       ter_temperature_measurement_at_triage  \
number                                         
97                              not_in_place   
96                                      <NA>   

       ter_physical_distancing_in_waiting  \
number                                      
97                           not_in_place   
96                                   <NA>   

       ter_separate_waiting_for_symptomatic chw_trained_precautions  \
number                                                                
97                             not_in_place            not_in_place   
96                                     <NA>                    <NA>   

       chw_trained_community_service     chw_drugs    chw_gloves chw_masks  \
number                                                                       
9

Where the rows are the same: 


facility_name region_province            district  chiefdom  \
number                                                               
97      RICHARD MCHP         WESTERN  WESTERN AREA URBAN  KENTUCKY   
96      RICHARD MCHP         WESTERN  WESTERN AREA URBAN  KENTUCKY   

       facility_location          facility_type facility_type_other  \
number                                                                
97          COLLEENBURGH  community_health_post                <NA>   
96          COLLEENBURGH  community_health_post                <NA>   

       managing_authority managing_authority_other setting  outpatient_only  \
number                                                                        
97                  other          private for fhc   urban             True   
96                  other          private for fhc   urban             True   

        capacity.number_consultation_rooms  capacity.number_inpatient_beds  \
number                                                                       
97                                       1                               6   
96                                       1                               6   

        capacity.number_maternity_beds  cho_number_in_post  \
number                                                       
97                                   2                   0   
96                                   2                   0   

        cho_number_present_at_visit  cht_number_in_post  \
number                                                    
97                                0                   0   
96                                0                   0   

        cht_number_present_at_visit  cha_number_in_post  \
number                                                    
97                                0                   0   
96                                0                   0   

        cha_number_present_at_visit  cm_number_in_post  sechn_number_in_post  \
number                                                                         
97                                0                  1                     1   
96                                0                  1                     1   

        sechn_number_present_at_visit  lab_tech_number_in_post  \
number                                                           
97                                  0                        0   
96                                  0                        0   

        lab_tech_number_present_at_visit  mch_aides_number_in_post  \
number                                                               
97                                     0                         4   
96                                     0                         4   

        mch_aides_number_present_at_visit  cleaner_porter_number_in_post  \
number                                                                     
97                                      0                              0   
96                                      0                              0   

        cleaner_porter_number_present_at_visit  other_number_in_post  \
number                                                                 
97                                           0                     1   
96                                           0                     1   

        other_number_present_at_visit other_staff_cadre  \
number                                                    
97                                  0      nursing aide   
96                                  0      nursing aide   

        consultations.head_count.month_2  consultations.head_count.month_3  \
number                                                                       
97                                  1266                              1079   
96                                  1266                              1079   

        consultations.head_count.month_4  \
number                           

Facility name: <NA>
Where the rows differ: 


facility_gps_lat  facility_gps_lon  consultations.head_count.month_1  \
number                                                                         
5              22.856753         19.262410                                23   
119            -1.129569        -34.179477                                56   

        consultations.head_count.month_2  consultations.head_count.month_3  \
number                                                                       
5                                     23                                23   
119                                    5                                53   

        consultations.head_count.month_4  \
number                                     
5                                     45   
119                                   34   

        consultations.general_outpatient.month_1  \
number                                             
5                                             85   
119                                           23   

        consultations.general_outpatient.month_2  \
number                                             
5                                             85   
119                                            3   

        consultations.general_outpatient.month_3  \
number                                             
5                                             84   
119                                           45   

        consultations.general_outpatient.month_4  \
number                                             
5                                             86   
119                                          656   

        consultations.deliveries.month_1  consultations.deliveries.month_2  \
number                                                                       
5                                     36                                35   
119                                   12                                 1   

        consultations.deliveries.month_3  consultations.deliveries.month_4  \
number                                                                       
5                                     12                                58   
119                                  415                                12   

        consultations.pent_vaccines.month_1  \
number                                        
5                                       120   
119                                      45   

        consultations.pent_vaccines.month_2  \
number                                        
5                                       125   
119                                       4   

        consultations.pent_vaccines.month_3  \
number                                        
5                                       639   
119                                      45   

        consultations.pent_vaccines.month_4       hr_focal_point  \
number                                                             
5                                       588  partially_completed   
119                                      45            completed   

       hr_staff_received_info iec_handwashing_procedure  \
number                                                    
5               not_completed             not_displayed   
119       partially_completed                 displayed   

       iec_early_symptom_recognition iec_when_facility_vs_home  \
number                                                           
5                          displayed             not_displayed   
119                    not_displayed                 displayed   

          surv_hotline_number surv_timely_data_reported_to_district  \
number                                                                
5       partially_operational                     fully_operational   
119         fully_operational                 partially_operational   

       chw_trained_precautions chw_trained_community_service chw_drugs  \
number                            

Where the rows are the same: 


facility_name region_province district chiefdom facility_location  \
number                                                                     
5               <NA>            <NA>     <NA>     <NA>              <NA>   
119             <NA>            <NA>     <NA>     <NA>              <NA>   

       facility_type facility_type_other managing_authority  \
number                                                        
5               <NA>                <NA>               <NA>   
119             <NA>                <NA>               <NA>   

       managing_authority_other setting  outpatient_only  \
number                                                     
5                          <NA>    <NA>             <NA>   
119                        <NA>    <NA>             <NA>   

        capacity.number_consultation_rooms  capacity.number_inpatient_beds  \
number                                                                       
5                                     <NA>                            <NA>   
119                                   <NA>                            <NA>   

        capacity.number_maternity_beds  cho_number_in_post  \
number                                                       
5                                 <NA>                <NA>   
119                               <NA>                <NA>   

        cho_number_present_at_visit  cht_number_in_post  \
number                                                    
5                              <NA>                <NA>   
119                            <NA>                <NA>   

        cht_number_present_at_visit  cha_number_in_post  \
number                                                    
5                              <NA>                <NA>   
119                            <NA>                <NA>   

        cha_number_present_at_visit  cm_number_in_post  \
number                                                   
5                              <NA>               <NA>   
119                            <NA>               <NA>   

        cm_number_present_at_visit  sechn_number_in_post  \
number                                                     
5                             <NA>                  <NA>   
119                           <NA>                  <NA>   

        sechn_number_present_at_visit  lab_tech_number_in_post  \
number                                                           
5                                <NA>                     <NA>   
119                              <NA>                     <NA>   

        lab_tech_number_present_at_visit  mch_aides_number_in_post  \
number                                                               
5                                   <NA>                      <NA>   
119                                 <NA>                      <NA>   

        mch_aides_number_present_at_visit  cleaner_porter_number_in_post  \
number                                                                     
5                                    <NA>                           <NA>   
119                                  <NA>                           <NA>   

        cleaner_porter_number_present_at_visit  other_number_in_post  \
number                                                                 
5                                         <NA>                  <NA>   
119                                       <NA>                  <NA>   

        other_number_present_at_visit other_staff_cadre  \
number                                                    
5                                <NA>               NaN   
119                              <NA>               NaN   

                                       sections_to_review  \
number                                                      
5       consultations hr information_education_communi...   
119     consultations hr information_education_communi...   

       hr_healthcare_provider_training  \
number                            

* JOHN CLINIC: The two rows are sufficiently different that I'm not sure enough that this is the same clinic to merge the rows into one or delete one of them. I therefore leave them as is. 
* KEVIN HOSPITAL: The two rows match in enough places that they are probably referring to the same clinic. They still differ in many places, however. I will merge them, but record both originals in a separate file and make the merging easy to comment out.  
* RICHARD MCHP: The two rows match in enough places that they are probably referring to the same clinic. They still differ in many places, however. I will merge them, but record both originals in a separate file and make the merging easy to comment out.  
* N/A: The rows are very different, so I leave these as two facilities. 


### Merging rows

I will merge the following rows:
* KEVIN HOSPITAL (indices 133 and 39)
* RICHARD MCHP (indices 96 and 97)
* All the facility name pairs in df_dup_2

Before doing any merging, I will export the duplicates to a csv file so that the machine learning engineer can examine it. 

In [283]:
df_dup = pd.concat([df_dup_2,mini_df_dict['KEVIN HOSPITAL'],mini_df_dict['RICHARD MCHP']])
print(f'Duplicates in df_dup: {sum(df_dup.index.duplicated())}')
df_dup.to_csv('similar_rows.csv')
df_dup

Duplicates in df_dup: 0


facility_name  facility_gps_lat  facility_gps_lon  \
number                                                                 
48                       AMY CHP         84.498725         48.271191   
87                       AMY CHP         70.462361        122.083912   
56                    GEORGE CHC        -85.563514       -123.262201   
110                   GEORGE CHC        -39.737294       -109.387845   
25                 JOANNA CLINIC         15.754015       -153.871776   
47                 JOANNA CLINIC         60.190934        162.787117   
86                MELISSA CLINIC         35.576733        179.184687   
113               MELISSA CLINIC        -29.047545       -115.874086   
115              NICOLE HOSPITAL         -1.046347        153.958550   
141              NICOLE HOSPITAL        -56.293473       -143.818347   
79      NICOLE TERTIARY HOSPITAL        -41.344766       -134.396481   
140     NICOLE TERTIARY HOSPITAL         63.155728        -82.989076   
83               ROBERT HOSPITAL         -6.673749       -118.231429   
16               ROBERT HOSPITAL         43.308952        134.454731   
101                SEAN HOSPITAL         81.489324        -62.510515   
128                SEAN HOSPITAL         11.498693        168.848917   
125                    TAMMY CHC        -11.824177         84.379063   
4                      TAMMY CHC         54.714883        150.744813   
133               KEVIN HOSPITAL         41.570610         66.484104   
39                KEVIN HOSPITAL        -53.597735        136.778861   
97                  RICHARD MCHP        -81.332628         69.862048   
96                  RICHARD MCHP        -26.341916       -172.661039   

       region_province            district       chiefdom facility_location  \
number                                                                        
48             WESTERN  WESTERN AREA URBAN      WISCONSIN    EAST JAMESBERG   
87             WESTERN  WESTERN AREA URBAN      WISCONSIN    EAST JAMESBERG   
56             EASTERN              KENEMA          TEXAS        PHILIPLAND   
110            EASTERN              KENEMA          TEXAS        PHILIPLAND   
25            SOUTHERN                  BO    MISSISSIPPI      AMBERCHESTER   
47            SOUTHERN                  BO    MISSISSIPPI      AMBERCHESTER   
86            NORTHERN           KOINADUGU      TENNESSEE     ANDREACHESTER   
113           NORTHERN           KOINADUGU      TENNESSEE     ANDREACHESTER   
115           SOUTHERN                  BO       VIRGINIA      STEVECHESTER   
141           SOUTHERN                  BO       VIRGINIA      STEVECHESTER   
79            SOUTHERN                  BO   RHODE ISLAND     NEW CHRISTINE   
140           SOUTHERN                  BO   RHODE ISLAND     NEW CHRISTINE   
83       NORTH-WESTERN           PORT LOKO       KENTUCKY    VALERIEBOROUGH   
16       NORTH-WESTERN           PORT LOKO       KENTUCKY    VALERIEBOROUGH   
101           SOUTHERN                  BO       NEW YORK        CLARKMOUTH   
128           SOUTHERN                  BO       NEW YORK         GUZMANTON   
125           SOUTHERN              BONTHE       VIRGINIA          RENEETON   
4             SOUTHERN              BONTHE       VIRGINIA          RENEETON   
133      NORTH-WESTERN              KARENE  MASSACHUSETTS        HUNTERPORT   
39       NORTH-WESTERN              KARENE  MASSACHUSETTS        HUNTERPORT   
97             WESTERN  WESTERN AREA URBAN       KENTUCKY      COLLEENBURGH   
96             WESTERN  WESTERN AREA URBAN       KENTUCKY      COLLEENBURGH   

                  facility_type facility_type_other managing_authority  \
number                                                                   
48        community_health_post                <NA>  government_public   
87        community_health_post                <NA>  government_public   
56      community_health_centre                <NA>  government_public   
110     community_healt

In [ ]:
df_amy_1 = df.loc[48]
df_amy_2 = df.loc[87]
df_amy_1

In [ ]:
df_amy_combo = df_amy_1.combine_first(df_amy_2)
df_amy_combo

In [ ]:
df = pd.concat([df, df_amy_combo.to_frame().T], ignore_index=True)
df = df.drop([48,87],axis=0)
df[df['facility_name']=='AMY CHP']

A number of facilities have the same name but are in different states. Since it is fairly unlikely that the state would be recorded incorrectly, these are likely cases where two independent facilities in different states happen to have the same name. Let's therefore require that the facility name and state are both the same and see if this reduces the number of duplicates. 

In [ ]:
#df[df.duplicated(['facility_name','facility_location'],keep=False)].sort_values('facility_location').to_csv('duplicates.csv')

In [ ]:
df[df.duplicated(['facility_name','chiefdom'],keep=False)].sort_values('facility_name')

It is possible that some duplicates are 'hiding' because the names are spelled slightly differently. Let's see if we can catch any like that by printing out all the facilities that have location matches with other facilities. 

In [ ]:
df[df.duplicated(['facility_location'],keep=False)].sort_values('facility_location')

From the above, it looks like there are no cases where facilities in the same location have two slightly different names. 

In [ ]:
print(df['form.health_centre_information.location_information.facility_location'].unique())

In [ ]:
df.to_csv('health_facility_assessment_cleaned.csv', header=True, index=True, index_label=None)